### ISSUE TITLE
Using WSL / Ubuntu on Windows 10

### ISSUE THREAD

**Issue by geerlingguy** [grasmash/drupal-composer-training#10]
I don't want to add anything to the official docs (at least not yet), but I do want to document here (in case I need it for reference) how to get everything working under WSL / Ubuntu on Windows 10 Pro:

  1. Install [PHP and Composer in Ubuntu using this guide](https://www.jeffgeerling.com/blog/2018/installing-php-7-and-composer-on-windows-10-using-ubuntu-wsl)
  1. Make sure you work within the shared C mount, e.g. inside `/mnt/c/Users/username/Sites` (this makes it easier to work with files between Ubuntu and Windows).
  1. If you need to call something like `docker-compose` installed with Docker for Windows, make sure you use `docker-compose.exe`.

Note that currently I get some errors if I try starting the Docker Compose environment within Ubuntu's CLI. If I switch over to PowerShell and run `docker-compose up -d` and manage things there, I can do so cleanly.

**Comment by geerlingguy** [comment]
Here's the error I'm getting:

```
PS C:\Users\jgeerling\Sites\my-project> docker-compose up -d
Creating drupal_vm_docker ... error

ERROR: for drupal_vm_docker  Cannot start service drupalvm: driver failed programming external connectivity on endpoint
drupal_vm_docker (5d63cdd47e4c5c15883515f16a82cb374d053f4b9c1637e6430888e2c89e06ad): Error starting userland proxy: mkdi
r /port/tcp:0.0.0.0:443:tcp:172.19.0.2:443: input/output error

ERROR: for drupalvm  Cannot start service drupalvm: driver failed programming external connectivity on endpoint drupal_v
m_docker (5d63cdd47e4c5c15883515f16a82cb374d053f4b9c1637e6430888e2c89e06ad): Error starting userland proxy: mkdir /port/
tcp:0.0.0.0:443:tcp:172.19.0.2:443: input/output error
ERROR: Encountered errors while bringing up the project.
```

I'm not entirely sure if this is a problem with docker-compose or something else local on my Windows 10 environment, though.

**Comment by geerlingguy** [comment]
Regarding the above errors: I just had to reset Docker to factory defaults in its settings and things seemed to start working again. Must've been some setting toggled by DDEV or Lando or something :/

**Comment by geerlingguy** [comment]
Works now! All from within WSL. So we can do it in WSL or native, I think that's a good spread of options.

![works](https://user-images.githubusercontent.com/481677/38626583-81206d48-3d72-11e8-8e86-17940e42e704.png)

**Comment by kslupsky** [comment]
I found the same error message trying to run a laradock installation on Windows Pro 10.

```
ERROR: for mysql  Cannot start service mysql: driver failed programming external connectivity on endpoint laradock_mysql_1 (d2f0c5a577e1f839beedb0b9d9cf7f0c157ab317c2b6b3189987dbaaea041c00): Error starting userland proxy: mkdir /port/tcp:0.0.0.0:3306:tcp:172.19.0.2:3306: input/output error

ERROR: for workspace  Cannot start service workspace: driver failed programming external connectivity on endpoint laradock_workspace_1
(38189f0be3c18c03850ba58a805f0d1e19acfa1cab6848a6d63733268e272d85): Error starting userland proxy: mkdir /port/tcp:0.0.0.0:2222:tcp:172.18.0.2:22: input/output error

ERROR: for nginx  Cannot start service nginx: driver failed programming external connectivity on endpoint laradock_nginx_1 (2db928fcafa1562c72a23bc03863f40f4cf00fd091422f4e27e3dc24cf005442): Error starting userland proxy: mkdir /port/tcp:0.0.0.0:443:tcp:172.19.0.2:443:
input/output error
```

I solved them by just restarting Docker. I changed no settings, didn't reset anything, I just had to restart. For some reason it seems like Docker maybe doesn't always start up correctly the first time it gets turned on.

**Comment by geerlingguy** [comment]
Yeah I've also seen some weird networking bugs get resolved by just restarting Docker for Windows, or restarting the entire Windows machine. There must be some part of the install process that doesn't set things correctly on the first run.

**Comment by haivx** [comment]
lau vkl
